### PR TITLE
fix: [WP-I6] add `onERC721Received` check for the `safeTransferFrom` functions of `LSP8CompatibleERC721` contracts

### PR DESCRIPTION
**Pull request by CJ42**
# What does this PR introduce?

## ⚠️ BREAKING CHANGES

In `LSP8CompatibleERC721` and `LSP8CompatibleERC721Init`:

- [x] make the `allowNonLSP1Recipient` parameter set to `true` by default for all transfer functions (`transfer` from LSP8, `transferFrom` and `safeTransferFrom` from ERC721).

- [x] for the `safeTransferFrom` function, add an extra call to check that the token receiver MUST implement the `onERC721Received` function and return the magic value (`onERC721Received(address,address,uint256,bytes)` function selector).

## 🧪 Tests

- [x] add tests for the following scenarios below.

- when `to` is an EOA = ✅ pass  
- when `to`  is a contract that supports + implement LSP1 
    - but does not implement `onERC721Received` = ❌ fail and revert  
    - if it also implements `onERC721Received`, but the function `onERC721Received` reverts = ❌ fails and revert  
    - if it also implements `onERC721Received` , but the function `onERC721Received`  does not return the right bytes value = ❌ fails and revert  
    - if it also implements `onERC721Received` = ✅ pass  

- when `to` is a contract that does not supports LSP1
    - but does not implement `onERC721Received` = ❌ fail and revert  
    - if it also implements `onERC721Received` , but the function `onERC721Received` reverts = ❌ fails and revert  
    - if it also implements `onERC721Received` , but the function `onERC721Received` does not return the right bytes value = ❌ fails and revert  
    - if it also implements `onERC721Received` = ✅ pass

### CODE REVIEW

File: tests/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721.test.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ describe("LSP8CompatibleERC721", () => {
     });
   });
 
-  describe("when using LSP8 contract with proxy", () => {
+  describe.skip("when using LSP8 contract with proxy", () => {
```
Comment (YamenMerhi) [inline tests/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721.test.ts:81]: `.skip`

File: contracts/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721InitAbstract.sol [hunk 2]
```diff
@@ -234,6 +256,42 @@ abstract contract LSP8CompatibleERC721InitAbstract is
...
+            try IERC721Receiver(to).onERC721Received(msg.sender, from, tokenId, data) returns (
+                bytes4 retval
+            ) {
+                return retval == IERC721Receiver.onERC721Received.selector;
+            } catch (bytes memory reason) {
+                if (reason.length == 0) {
+                    revert("LSP8CompatibleERC721: transfer to non ERC721Receiver implementer");
+                } else {
+                    /// @solidity memory-safe-assembly
+                    assembly {
```
Comment (YamenMerhi) [inline contracts/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721InitAbstract.sol:285]: linter error for assembly

File: contracts/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721.sol [hunk 3]
```diff
@@ -244,6 +260,42 @@ abstract contract LSP8CompatibleERC721 is
...
+            try IERC721Receiver(to).onERC721Received(msg.sender, from, tokenId, data) returns (
+                bytes4 retval
+            ) {
+                return retval == IERC721Receiver.onERC721Received.selector;
+            } catch (bytes memory reason) {
+                if (reason.length == 0) {
+                    revert("LSP8CompatibleERC721: transfer to non ERC721Receiver implementer");
+                } else {
+                    /// @solidity memory-safe-assembly
+                    assembly {
```
Comment (YamenMerhi) [inline contracts/LSP8IdentifiableDigitalAsset/extensions/LSP8CompatibleERC721.sol:289]: linter error for assembly

**Review by YamenMerhi — CHANGES_REQUESTED** [review]
First round of review

**Review by skimaharvey — APPROVED** [review]

**Review by b00ste — APPROVED** [review]

**Review by YamenMerhi — APPROVED** [review]
LGTM 🚀 What would be a good additional test is to test whether using `transferFrom` or `safeTransferFrom` when `onERC721Received(..)` pass and LSP1 revert, the whole call will revert even if `allowNonLSP1Recipient` is true.

**Pull request merged**